pkg/costmodel/allocation.go

```diff
@@ -720,302 +786,422 @@ func applyPodResults(window kubecost.Window, resolution time.Duration, podMap ma
+				log.DedupedWarningf(10, "CostModel.ComputeAllocation: CPU allocation query result missing 'container': %s", key)
+				continue
+			}
 
-		if _, ok := pod.Allocations[container]; !ok {
-			pod.AppendContainer(container)
-		}
+			if _, ok := pod.Allocations[container]; !ok {
+				pod.AppendContainer(container)
+			}
 
-		cpuCores := res.Values[0].Value
-		if cpuCores > MAX_CPU_CAP {
-			log.Infof("[WARNING] Very large cpu allocation, clamping to %f", res.Values[0].Value*(pod.Allocations[container].Minutes()/60.0))
-			cpuCores = 0.0
-		}
-		hours := pod.Allocations[container].Minutes() / 60.0
-		pod.Allocations[container].CPUCoreHours = cpuCores * hours
+			cpuCores := res.Values[0].Value
+			if cpuCores > MAX_CPU_CAP {
+				log.Infof("[WARNING] Very large cpu allocation, clamping to %f", res.Values[0].Value*(pod.Allocations[container].Minutes()/60.0))
+				cpuCores = 0.0
+			}
+			hours := pod.Allocations[container].Minutes() / 60.0
+			pod.Allocations[container].CPUCoreHours = cpuCores * hours
 
-		node, err := res.GetString("node")
-		if err != nil {
-			log.Warningf("CostModel.ComputeAllocation: CPU allocation query result missing 'node': %s", key)
-			continue
+			node, err := res.GetString("node")
+			if err != nil {
+				log.Warningf("CostModel.ComputeAllocation: CPU allocation query result missing 'node': %s", key)
+				continue
+			}
+			pod.Allocations[container].Properties.Node = node
 		}
-		pod.Allocations[container].Properties.Node = node
 	}
 }
 
-func applyCPUCoresRequested(podMap map[podKey]*Pod, resCPUCoresRequested []*prom.QueryResult) {
+func applyCPUCoresRequested(podMap map[podKey]*Pod, resCPUCoresRequested []*prom.QueryResult, podUIDKeyMap map[podKey][]podKey) {
 	for _, res := range resCPUCoresRequested {
 		key, err := resultPodKey(res, env.GetPromClusterLabel(), "namespace")
 		if err != nil {
 			log.DedupedWarningf(10, "CostModel.ComputeAllocation: CPU request result missing field: %s", err)
 			continue
 		}
 
+		var pods []*Pod
+
 		pod, ok := podMap[key]
 		if !ok {
-			continue
-		}
-
-		container, err := res.GetString("container")
-		if err != nil {
-			log.DedupedWarningf(10, "CostModel.ComputeAllocation: CPU request query result missing 'container': %s", key)
-			continue
+			if uidKeys, ok := podUIDKeyMap[key]; ok {
+				for _, uidKey := range uidKeys {
+					pod, ok = podMap[uidKey]
+					if ok {
+						pods = append(pods, pod)
+					}
+				}
+			} else {
+				continue
+			}
+		} else {
+			pods = []*Pod{pod}
 		}
 
-		if _, ok := pod.Allocations[container]; !ok {
-			pod.AppendContainer(container)
-		}
+		for _, pod := range pods {
+			container, err := res.GetString("container")
+			if err != nil {
+				log.DedupedWarningf(10, "CostModel.ComputeAllocation: CPU request query result missing 'container': %s", key)
+				continue
+			}
 
-		pod.Allocations[container].CPUCoreRequestAverage = res.Values[0].Value
+			if _, ok := pod.Allocations[container]; !ok {
+				pod.AppendContainer(container)
+			}
```

Comment:
Say we have the following situation: two pods, with the same name, in the same window, but with different containers running. Like so:

![same-pod-different-container-cpu-request](https://user-images.githubusercontent.com/17585786/161841348-035d3dec-6235-465d-83a0-c0f51d1bbdc9.png)

What kind of `Allocation`s do we end up with?

My suspicion is that we get 4:
1) Pod name: "A 1" +
    Container name "C1" + 
    CPUCoreRequestAverage: 1 +
    Minutes: ~half of window
2) Pod name: "A 1" +
    Container name "C2" + 
    CPUCoreRequestAverage: 3 +
    Minutes: ~half of window
3) Pod name: "A 2" +
    Container name "C1" + 
    CPUCoreRequestAverage: 1 +
    Minutes: ~half of window
4) Pod name: "A 2" +
    Container name "C2" + 
    CPUCoreRequestAverage: 3 +
    Minutes: ~half of window

In this situation, do we double-count the cost of "UID 1 C1" and "UID 2 C2"?

(I'm picking this line of code merely for convenience, this is a general design question about applying metric information to _all_ pods with the same name if the UID is not available for a given metric)

Reply:
Kaelan and I discussed. One thing I'd like to note is that according to https://github.com/kubecost/cost-model/blob/485851c6fbd8f1d23656bb279cce38006ffaa0a8/pkg/costmodel/allocation.go#L2522-L2538, the container start and end come from the POD start and end. So we're already wrong in this case right now.

Reply:
As Michael says, this is extant behavior. In the above case currently (or with these UID changes off), the pod will simply show one pod named "A" with two containers "C1" and "C2" running for its entire runtime.

This is effectively the same double-counting as which Michael explains above: two pods "A 1" and "A 2" having two containers each running for each pod's entire runtime is, added up, approximately the same runtime as the one pod "A" with two containers, just more "visibly wrong" because we can see the different pods but the containers are incorrect.

Discussed this with Ajay. There's really no way to mitigate this, the only link between how we get the containers/allocations and the pods is the name. We can't see which container belongs to which pod specifically if 1. the pod names are the same, and 2. the container names are different (and therefore a distinct entry in pod.Allocations), so this is just an unfortunate edge case.